### ISSUE TITLE
Unified point + text rendering style

### ISSUE THREAD

**Issue by bcamper** [tangrams/tangram#290]
Currently, `points` (which can be rendered by shaders or optionally configured to draw a sprite) and `text` are two separate rendering styles. This works fine for standalone labels (e.g. roads) or icons, but presents a challenge for typical composite point/text rendering combinations, such as POI icons:

- **Styling is cumbersome**, especially because the point and text components are two independent entities that have no knowledge of each other or logic controlling their rendering in tandem.
   - For example, text must be manually placed relative to the point: to apply a bottom anchor to a text label appearing below an icon, an additional `offset` must be provided to move the text downwards half the height of the icon; setting `anchor: bottom` on the text is also required (as expected), but only positions the text such that its top is directly below the geometric feature point, without taking into account any space occupied by the icon itself.
   - This is further exacerbated in cases where the icon size itself changes across zoom levels; the text offset must also track this size! Here's an abridged example from Bubble Wrap:
   ```
   draw:
      icons:
         size: [[13, 12px], [16, 16px], [18, 19px]]
      text:
         anchor: bottom
         offset: [[13, [0, 6px]], [16, [0, 8px]], [18, [0, 10px]]] # offset tracks half icon size
   ```
- **Placement logic is limited** due to the lack of direct link between the two components. For example, we cannot currently make a text label only display when its linked icon is visible (we can set the `priority` of the icon to be higher than the text, but this does not guarantee that they will display together).

Proposal
===
- The most straightforward way to create a unified rendering style that allows for styling points and text in tandem is to extend the existing `points` style to have an *optional* text component. 
- This text component would appear as a `text` block within a `points` draw rule (similar to the `outline` block within a `lines` draw rule). For example, to create an icon with a text label, using a style `icons` that has `base: points`:

   ```
   draw:
      icons:
         sprite: ...
         size: 16px
         text:
            font: ...
   ```
- A `text` block attached to a `points`-based style would have slightly different default values (again, similar to outlines):
   - **Anchor**:
      - `anchor` would default to `bottom` instead of `center` (though it would be possible to composite a text label over a sprite by setting `anchor: center` and `collide: false`)
      - The point and text can have separate `anchor` values: the `anchor` of the `text` controls the text's placement *relative to the size and position* of its "parent" point, while the `anchor` of the `points` portion moves the *entire entity* (point + text) relative to the underlying feature's point.
   - **Offset**:
      - An additional offset adjustment would be automatically applied so that the text anchor is truly relative to its "parent" point (see the layout problem described above).
      - Further manual offset would still be possible with the `offset` parameter, which moves the text in screen space, e.g. text with `anchor: bottom` will automatically be placed below the sprite, and an additional `offset: [8px, 0]` in the scene file would move the text another 8 pixels to the right.
   - **Priority**:
      - The text's `priority` would have a default value of `0.5` below that of the `priority` specified for its parent. This could be explicitly overridden by setting a `priority` value in the `text` block, though the text's priority could not be set higher than that of its parent point. This is similar to `outline` handling, where the `order` of the outline cannot be higher than the line fill.
- **Collision**:
   - While the `points` portion of the style would render according to its collision test, the `text` portion would only render if both it and its "parent" point passed collision tests, e.g. if the point is occluded, then the text won't render either, even if it is not occluded.
- **Migration/Compatibility**:
   - The `text` style would continue to exist, for cases where only a text label is needed (road labels are a clear example). With additive syntax on `points`, we can maintain full backwards compatibility.
   - The migration path is also straightforward: in cases where an existing `points`/`text` pair already exists, the `text` portion can simply be indented under `points`, and parameters adjusted (for instance, removing custom `offsets`).

**Comment by nvkelso** [comment]
For collision tests, would the priority (`p`) of the example `icon` (label) block alone suffice to imply the correct say priority `p - 0.5`for the child text?

Offset for text from sprite is still useful, but would it be measured from the edge of the sprite now instead of the center of the point?

**Comment by bcamper** [comment]
@nvkelso yes that's what I was thinking on priority: default text priority below the sprite's priority, and allow an override (this follows our outline behavior, where outline order defaults to a half-value below the line fill, but can be overridden for example to group outlines across several road classes under the same order).

For `offset`, yes, an additional offset would still be possible. But offset is already applied in screen space, so it would function the same regardless of its positioning relative to a sprite, e.g. if text had `anchor: bottom`, that moves the text below the sprite, and the text could apply an additional `offset: [8px, 0]` to move itself 8 pixels to the right.

In fact, the text portion would have the `anchor` and `offset` logic applied for *both* itself and its "parent" sprite.

**Comment by bcamper** [comment]
@nvkelso updated description to reflect comments

**Comment by bcamper** [comment]
WIP here: https://github.com/tangrams/tangram/compare/unify-points-text

**Comment by bcamper** [comment]
@karimnaaji could you take a look at this both from a proposed syntax proposal (nesting a `text` block inside a `points` draw block, like we do with `outline` under `line`), and also any thoughts JS implementation wise on this branch: https://github.com/tangrams/tangram/compare/unify-points-text

Right now I'm using the technique we discussed where the points/sprites and text labels are all in the same mesh, with two textures bound, and an attribute/varying used to [branch between the two samplers](https://github.com/tangrams/tangram/blob/bcb015b2121bc8667cbbe188bf2bd259d370a461/src/styles/points/points_fragment.glsl#L47-L65). This is the most expedient way to get this working with the current one-draw-call-per-style pipeline, but could be pretty easily split into two buffers + draw calls in the future without otherwise changing the logic.

**Comment by karimnaaji** [comment]
This looks pretty consistent to me; the way offsets adjusts to their parent is mostly what I was thinking offsets when suggesting _automatic_ offsets for compound elements a while back, from a stylesheet point of view.

> This could be explicitly overridden by setting a priority value in the text block, though the text's priority could not be set higher than that of its parent point. This is similar to outline handling, where the order of the outline cannot be higher than the line fill.

I think it's a bit different than polyline outlines, since here we're displaying information. We should let the user specify higher priority for the text if this is something desired. I understand the reasons of having the text priority 0.5 lower than its parent by default, but I feel like locking priorities to either one of the parents/child could be limiting. 

But the question is, do we want the user to always use the unified style? or, are we still going to support sprite/text as two separate styles?

If we don't want to support styles as separate entities, cases where this is limiting would be where you want all sprites to render (collision to false) but text to be colliding between each other. Here we would be narrowing down our case to small icons + text, which is would be most of the use cases, but I can see having sprite icons size changing dynamically based on a feature property (population, etc) where sprite size would be a dynamic size; it's not a feature we have right now but that I have in mind for a more data visualization use case.

On -es, we don't use the same shader/vertex attributes to render text/sprite so this would be a different implementation (one is using distance field, where the other doesn't) so we will use different meshes; which will also be easier considering our texture paging for multiple atlases involved by the text rendering.

**Comment by bcamper** [comment]
My motivations are trying to balance the expected common use cases, backwards compatibility, and functionality/code complexity, which led me to:

- **Preserving existing** rendering style syntax for `points` and `text` (e.g. they are fully backwards compatible; this isn't necessarily a hard requirement, but is the easiest migration path follows naturally from the rest of this proposal).
- **No new** rendering styles needing to be added (e.g. just using `points` and `text`).
- **Adding new functionality** to the `points` style to support text labels that pair with a point (could be a sprite or just a plain shader-rendered point).
  - As proposed, in the initial implementation the point would be **required** while the text would be **optional** (e.g. point will render without text, but text won't render without point). This follows my understanding of the most common cartographic use case (POIs, often seeing the icon at lower zooms, with more labels being added as you zoom in). The text is secondary to the point, which provides grounding. If you aren't so concerned with these things being connected, you can use the existing styles independently as we do now.
  - You can still set the `collide: false` flag on either or both of the point and text independently:
    - Both `collide: true` (default): nothing will overlap, text will only be rendered if point also fits.
    - Points `collide: false`, text `collide: true`: all points will render, text will render if it fits.
    - Points `collide: true`, text `collide: false`: points will render if they fit, in which case their attached text will also render, even if it overlaps something else.
    - Both `collide: false`: all points and text should render, regardless of overlap.
    - Note: in JS there's some additional logic for culling objects that intersect tile bounds, that's easier discussed separately.
  - Without an attached `text` block, the `points` style continues as a standalone points-only style as it is today.
  - I believe this provides a wide range of use cases initially, but agree we can potentially expand with flags to support others if there's sufficient demand.
- **Continuing** to use the `text` **standalone** style for case where you **only** need to render text (e.g. no icon). Most commonly this means road labels, which also have additional text layout logic for angle and in the future curves (these are not needed for point + text compounds), but it could also be used for text-only point labels.

**Comment by nvkelso** [comment]
We also use stand alone (no icon) text labels on buildings in all the house
styles.

**Comment by bcamper** [comment]
@nvkelso yeah, and my goal is to continue to support those without needing any syntax changes to existing styles.

**Comment by karimnaaji** [comment]
@bcamper thanks for clarifying. This makes sense if support for standalone style is still available.

**Comment by bcamper** [comment]
Shipping as part of 0.8.

**Comment by bcamper** [comment]
Released in [v0.8.0](https://github.com/tangrams/tangram/releases/tag/v0.8.0)
